internal/router/router.go

```diff
@@ -130,9 +128,59 @@ type Handlers struct {
 type Router struct {
 	handlers   *Handlers
 	handler    http.Handler
-	resolveURL func(id *charm.Reference) error
-	authorize  func(id *charm.Reference, req *http.Request) error
-	exists     func(id *charm.Reference, req *http.Request) (bool, error)
+	resolveURL func(id *charm.Reference) (*ResolvedURL, error)
+	authorize  func(id *ResolvedURL, req *http.Request) error
+	exists     func(id *ResolvedURL, req *http.Request) (bool, error)
+}
+
+// ResolvedURL represents a URL that has been resolved by resolveURL.
+// URL.User and URL.Series should always be non-empty and
+// URL.Revision should never be -1.
+//
+// If PromulgatedRevision is not -1, it holds the revision of the
+// promulgated version of the charm.
+type ResolvedURL struct {
+	URL                 charm.Reference
+	PromulgatedRevision int
+}
+
+// MustNewResolvedURL returns a new ResolvedURL by parsing
+// the entity URL in urlStr. The promulgatedRev parameter
+// specifies the value of PromulgatedRevision in the returned
+// value.
+//
+// This function panics if urlStr cannot be parsed as a charm.Reference
+// or if it is not fully specified, including user, series and revision.
+func MustNewResolvedURL(urlStr string, promulgatedRev int) *ResolvedURL {
+	url := charm.MustParseReference(urlStr)
+	if url.User == "" || url.Series == "" || url.Revision == -1 {
+		panic(fmt.Errorf("incomplete url %v", urlStr))
+	}
+	return &ResolvedURL{
+		URL:                 *url,
+		PromulgatedRevision: promulgatedRev,
+	}
+}
+
+// PreferredURL returns the promulgated URL for
+// the given id if there is one, otherwise it
+// returns &id.Reference. The returned *charm.Reference
+// may be modified freely.
+func (id *ResolvedURL) PreferredURL() *charm.Reference {
+	if id.PromulgatedRevision == -1 {
+		u := id.URL
```

Comment:
We could assign u := id.URL before the if block and use it for both cases [shrug].